caffe2/contrib/tensorrt/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,195 @@
+    // We conform to the output shape hints. NB: We might need an explicit reshape op for this
+    if (!identical_shape) {
+      *dims = dims_hint;
+    }
+  }
+}
+
+bool TensorRTOp::RunOnDevice() {
+  CAFFE_ENFORCE(trt_executor_);
+  // Decide input batch size
+  size_t N = 0;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    if (i == 0 && !tensor_dims.empty()) {
+      N = tensor_dims.front();
+    } else {
+      CAFFE_ENFORCE_EQ(
+          N, tensor_dims.front(), "Mismatched batch size in input tensors");
+    }
+  }
+  if (N > max_batch_size_) {
+    LOG(WARNING) << "Batch size (" << N << ") is larger than max_batch_size ("
+                 << max_batch_size_ << ") optimized for TensorRT operator. "
+                 << "Performance may be sub-optimal.";
+  }
+
+  // We need to do the binding at RunOnDevice time because we only know the
+  // exact shapes of the tensors now. In addtion, since TensorRT engine has
+  // max_batch_size, we need to call that multiple times if input batch size
+  // exceeeds this limit.
+  CAFFE_ENFORCE_EQ(is_input_.size(), nv_dims_.size());
+  std::vector<void*> bindings;
+  bindings.reserve(is_input_.size());
+  auto batch_size = max_batch_size_;
+  for (size_t offset = 0; offset < N; offset += batch_size) {
+    bindings.clear();
+    batch_size =
+        offset + max_batch_size_ > N ? N - offset : max_batch_size_;
+    VLOG(2) << "Offset: " << offset << ", batch_size: " << batch_size << ", N: " << N;
+    int input_idx = 0;
+    int output_idx = 0;
+    for (auto i = 0; i < is_input_.size(); ++i) {
+      const auto& dims = nv_dims_[i];
+      if (is_input_[i]) {
+        // input, check input dimensions
+        const auto& input_tensor = Input(input_idx++);
+        const float* input_data = input_tensor.data<float>();
+        const auto& tensor_dims = input_tensor.dims();
+        auto chw = CheckDims(dims, tensor_dims);
+        bindings.push_back((void*)(input_data + offset * chw));
+      } else {
+        // output, we need to allocate the output tensor at first batch run
+        auto* output_tensor = Output(output_idx);
+        std::vector<TIndex> tensor_dims;
+        tensor_dims.push_back(N);
+        int64_t chw = 1;
+        for (int i = 0; i < dims.nbDims; ++i) {
+          tensor_dims.push_back(dims.d[i]);
+          chw *= dims.d[i];
+        }
+
+        if (offset == 0) {
+          MaybeAdjustOutputShape(output_idx, &tensor_dims);
+          output_tensor->Resize(tensor_dims);
+        }
+        ++output_idx;
+        float* output_data = output_tensor->mutable_data<float>();
+        bindings.push_back((void*)(output_data + offset * chw));
+      }
+    }
+
+    CAFFE_ENFORCE_EQ(bindings.size(), InputSize() + OutputSize());
+    if(!trt_executor_->execute(batch_size, bindings.data())){
+      CAFFE_THROW("Error running the TensorRT executor");
```

Comment:
does TRT have some reasonable error string?

Reply:
It prints out info from Ilogger interface which we already provided.